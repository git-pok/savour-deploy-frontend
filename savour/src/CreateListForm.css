.CreateListForm-form {
  display: flex;
  flex-direction: row;
  justify-content: center;
  flex-wrap: wrap;
  width: 70%;
  margin: 20px auto 40px auto;
  padding: 30px 0px;
  /* background-color: rgb(39, 55, 77); */
  background-color: rgb(78, 42, 42);
  border-radius: 10px;
}

.CreateListForm-form-field {
  display: block;
  width: 46%;
  margin: 10px 10px;
}

.CreateListForm-form-field label {
  display: block;
  width: 100%;
  height: 30px;
  margin: 0px auto;
  /* color: rgb(221, 230, 237); */
  color: rgb(255, 255, 255);
}

.CreateListForm-form-field input, .CreateListForm-form-field select {
  box-sizing: border-box;
  width: 100%;
  height: 40px;
  padding: 0px 0px;
  margin: 0px auto 20px auto;
  border: 6px solid rgb(255, 255, 255);
  border-radius: 05px;
}

.CreateListForm-form-submit {
  display: block;
  width: 100%;
  margin-top: 10px;
}

.CreateListForm-form-submit button {
  display: block;
  width: 260px;
  height: 30px;
  margin: 0px auto;
  margin-top: 10px;
  border: 3px inset rgb(223, 216, 85);
  background-color: rgb(223, 216, 85);
  /* background-color: rgb(221, 230, 237); */
  color: rgb(0, 0, 0);
  transition: .5s ease-in-out background-color;
}

.CreateListForm-form button:hover {
  border-color: rgb(255, 255, 255);
  background-color: rgb(255, 255, 255);
  color: rgb(0, 0, 0);
}

.CreateListForm-msg-div {
  display: block;
  width: 50%;
  margin: 0px auto;
  padding: 0px 0px;
}

.CreateListForm-msg-div div {
  margin: 0px auto;
}

@media only screen and (max-width: 1384px) and (min-width: 400px) {
  .CreateListForm-div {
    flex-direction: column;
    align-items: center;
    flex-wrap: nowrap;
  }

  .CreateListForm-form-field {
    width: 70%;
  }

  .CreateListForm-form-submit button {
    width: 50%;
    /* height: 100%; */
  }
}

