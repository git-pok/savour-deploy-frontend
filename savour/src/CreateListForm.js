import { useContext, useState, useEffect } from 'react';
import useAxios from './hooks/useAxios.js';
import UserContext from './context/UserContext.js';
import SavourApi from './models/SavourApi.js';
import useToggleState from './hooks/useToggleState.js';
import Message from './Message.js';
import './CreateListForm.css';
/**
 * CreateListForm
 * Props: recipeId, setState, recipelist
 * Renders form to create shop or recipe list.
*/
const CreateListForm = ({recipeId, setState, recipelist=false}) => {
  const { usrData, setUsrData } = useContext(UserContext);
  const headers = { _token: `Bearer ${usrData.token}`};

  const recipeOpts = { method: "get", url: `/recipes`, data: {}, params: {}, headers };
  const [ recipeData ] = useAxios(recipeOpts);

  const OccOpts = { method: "get", url: `/occasions`, data: {}, params: {}, headers };
  const [ occasionData ] = useAxios(OccOpts);

  const [ listUrl, setListUrl ] = useState(null);
  const [ isRecipeList, setIsRecipeList ] = useToggleState(false);
  const [ isSubmitted, setIsSubmitted ] = useToggleState(false);

  const initialState = {
    shoplistName: "", recipelistName: "", occasionId: ""
  };

  const [ formData, setFormData ] = useState(initialState);
  const [ formReqMade, setFormReqMade ] = useToggleState(false);
  const [ formErrMsg, setFormErrMsg ] = useState(null);
  const [ succMsg, setSuccMsg ] = useState(null);
  const [ isFormReqSucc, setIsFormReqSucc ] = useToggleState(false);

  const setReqUrl = (shop=true) => {
    const userId = usrData.userId;
    // Check for occasion id from recipelist form data.
    const occId = formData.occasionId;
    if (!shop && occId === "") {
      setFormErrMsg("Must select occasion!");
      setFormReqMade();
      setTimeout(setFormReqMade, 3000);
      setTimeout(() => setFormErrMsg(null), 3000);
    } else {
      if (!shop) setListUrl(`/users/${userId}/recipelists`);
      if (!shop) setIsRecipeList();
      else setListUrl(`/users/${userId}/shoppinglists`);
      setIsSubmitted();
    }
  }

  useEffect(() => {
    const createList = async () => {
      try {

        const data = isRecipeList ?
            { occasion_id: +formData.occasionId, list_name: formData.recipelistName }
            : { recipe_id: +recipeId, list_name: formData.shoplistName }

        const listReq = await SavourApi.request("post", listUrl, data, {}, headers);
        setSuccMsg("Created list!");
        setFormReqMade();
        setIsFormReqSucc();
        setTimeout(setIsSubmitted, 3000);
        if (isRecipeList) setTimeout(setIsRecipeList, 3000);
        setTimeout(setFormReqMade, 3000);
        setTimeout(setIsFormReqSucc, 3000);
        setTimeout(() => setSuccMsg(null), 3000);
        setTimeout(() => setListUrl(null), 3000);

        const listReReq = await SavourApi.request("get", listUrl, {}, {}, headers);
        if (setState) setTimeout(() => setState(() => [...listReReq.data]), 3000);
      } catch(err) {
        console.log("ERROR", err);
        const error = err.response.data.error.message;
        const isErrObj = typeof err.message === "object";
        const isErrorObj = typeof error === "object";
        const errMsg = isErrObj ? null : err.message;
        const errorMsg = isErrorObj ? null : error;
        setFormErrMsg(() => errorMsg || errMsg || "Error!");
        setFormReqMade();
        setTimeout(setIsSubmitted, 3000);
        if (isRecipeList) setTimeout(setIsRecipeList, 3000);
        setTimeout(setFormReqMade, 3000);
        setTimeout(() => setFormErrMsg(null), 3000);
        setTimeout(() => setListUrl(null), 3000);
      }
    }

    if (isSubmitted) createList();
  }, [isSubmitted])

  const handleChange = (evt) => {
    const { name, value } = evt.target;

    setFormData(data => ({
      ...data,
      [name]: value
    }))
  }

  const handleSubmit = async (evt) => {
    evt.preventDefault();
  }

  return (
    <>
    {/* Shoppinglist form to create shoppinglist. */}
    { !recipelist
      ?
        <form onSubmit={handleSubmit} className="CreateListForm-form">
          <div className="CreateListForm-form-field">
            <label htmlFor="shoplistName">List Name</label>
            <input
              id="shoplistName"
              type="text"
              name="shoplistName"
              placeholder="Type a list name"
              onChange={handleChange}
              value={formData.shoplistName} />
          </div>
          <div className="CreateListForm-form-submit">
          <div className="CreateListForm-msg-div">
            { formReqMade &&
              <Message msgObj={
                {
                  class: isFormReqSucc ? "success" : "fail",
                  msg: isFormReqSucc ? succMsg : formErrMsg
                }
              } />
            }
          </div>
            <button onClick={setReqUrl}>CREATE LIST</button>
          </div>
        </form>
      :
        // {/* Recipelist form to create recipelist. */}
        <form onSubmit={handleSubmit} className="CreateListForm-form">
          <div className="CreateListForm-form-field">
            <label htmlFor="recipelistName">List Name</label>
            <input
              id="recipelistName"
              type="text"
              name="recipelistName"
              placeholder="Type a recipelist name"
              onChange={handleChange}
              value={formData.recipelistName}></input>
          </div>
          <div className="CreateListForm-form-field">
            <label htmlFor="occasionId">Recipelist Occasions</label>
            <select
              id="occasionId"
              name="occasionId"
              onChange={handleChange}
              value={formData.occasionId}>
            <option key="select-an-occ" value="">Select an Occasion</option>
              { occasionData &&
                occasionData.map(occasion => (
                  <option key={`${occasion.id}`} value={`${occasion.id}`}>{occasion.occasion}</option>
                ))
              }
            </select>
          </div>
          <div className="CreateListForm-form-submit">
          <div className="CreateListForm-msg-div">
            { formReqMade &&
              <Message msgObj={
                {
                  class: isFormReqSucc ? "success" : "fail",
                  msg: isFormReqSucc ? succMsg : formErrMsg
                }
              } />
            }
          </div>
            <button onClick={() => setReqUrl(false)}>CREATE LIST</button>
          </div>
        </form>
    }
    </>
  );
}

export default CreateListForm;