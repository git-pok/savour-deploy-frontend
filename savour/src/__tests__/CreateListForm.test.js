import CreateListForm from '../CreateListForm.js';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import UserContext from '../context/UserContext.js';


const usrData = {
  userUsername: "fvin",
  userId: 12
};

const setState = () => {
  console.log("Test!");
}

test("renders CreateListForm without crashing", () => {
  render(
    <MemoryRouter>
      <UserContext.Provider value={{usrData}}>
        <CreateListForm
          recipeId={12}
          setState={setState} />
      </UserContext.Provider>
    </MemoryRouter>
  );
});

test("renders snapshot without crashing", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <UserContext.Provider value={{usrData}}>
        <CreateListForm
          recipeId={12}
          setState={setState} />
      </UserContext.Provider>
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});