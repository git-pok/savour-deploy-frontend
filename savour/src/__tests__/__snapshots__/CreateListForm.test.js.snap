// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot without crashing 1`] = `
<DocumentFragment>
  <form
    class="CreateListForm-form"
  >
    <div
      class="CreateListForm-form-field"
    >
      <label
        for="shoplistName"
      >
        List Name
      </label>
      <input
        id="shoplistName"
        name="shoplistName"
        placeholder="Type a list name"
        type="text"
        value=""
      />
    </div>
    <div
      class="CreateListForm-form-submit"
    >
      <div
        class="CreateListForm-msg-div"
      />
      <button>
        CREATE LIST
      </button>
    </div>
  </form>
</DocumentFragment>
`;
